### DISCUSSION TITLE
Can't get rolls to work.

### DISCUSSION THREAD

**Discussion by sommerper** [getreuer/qmk-keymap#21] [Q&A] [answered]
Hi!

I've implemented the code like so.

```
void matrix_scan_user(void) {
  achordion_task();
}

bool process_record_user(uint16_t keycode, keyrecord_t *record) {
   if (!process_achordion(keycode, record)) { return false; }
    return true;
}
```

I've got the left homerow on ARST (colemak).
If I roll over R and S I usually only get the S. I can't figure out how to implement the code properly even though I'm trying to follow the guide on https://getreuer.info/posts/keyboards/achordion/index.html

I'd be happy if somebody could give me some hints on how to achieve that.
Thanks!

**Comment by getreuer** [comment, marked as answer]
Thanks for the report! I take it this is home row mods and both `R` and `S` are mod-tap keys? A roll over `RS` is intended to work.

Do you by chance have Combos involving either `R` or `S`? Combos are a tricky aspect of Achordion's implementation, so it's the first thing to come to mind.

What tap-hold configuration options do you use? You could try [what I use](https://getreuer.info/posts/keyboards/achordion/index.html#tap-hold-configuration) and see whether that makes a difference.

If you enable debug logging [as described here](https://docs.qmk.fm/#/faq_debug), Achordion will print messages when something interesting happens, e.g. Achordion logs "`Achordion: Plumbing hold press`" when a key has been settled as a hold. This might give some insight about how the `RS` roll is getting mishandled.

Another thought: while Achordion is intended to address this roll, you could (with or without Achordion) try [precondition's Rolled modifiers cancellation](https://precondition.github.io/home-row-mods#rolled-modifiers-cancellation). The idea is: if `S` is tapped while `R` is held, then revise both keys as held.

**Comment by sommerper** [comment]
Thanks a bunch! I had understood that part incorrectly as I thought that was about exceptions. Now it works beautifully. Much better than my own implementation.

I did have to add the layer mods like in the function below so I wouldn't have to wait for the timeout.
```
bool achordion_chord(uint16_t tap_hold_keycode,
                     keyrecord_t* tap_hold_record,
                     uint16_t other_keycode,
                     keyrecord_t* other_record) {

  switch (tap_hold_keycode) {
    case TGL_L_BKSP:
      return true;
      break;
    case TGL_R_SPC:
      return true;
      break;
  }
  return achordion_opposite_hands(tap_hold_record, other_record);
}
```

I can't live without this on my Sweep!

Thanks!

**Reply by getreuer** [reply]
Excellent, glad you worked it out!

Yes, what you did with `TGL_L_BKSP` and `TGL_R_SPC` in `achordion_chord()` makes sense. If you like, you can have a certain tap-hold key bypass Achordion completely by having `achordion_timeout()` return 0, as [described here](https://getreuer.info/posts/keyboards/achordion/index.html#achordion_timeout).
